### PR TITLE
new(falco): add gVisor support

### PR DESCRIPTION
**Pull request by LucaGuerra**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

> If contributing rules or changes to rules, please make sure to also uncomment one of the following line:

> /kind rule-update

> /kind rule-create

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area rules

> /area tests

> /area proposals

**What this PR does / why we need it**:

This PR adds support for the Falco CLI to enable syscall data collection from gVisor, as supported by libsinsp and libscap. Before this can be tested it needs the libs version to be advanced to the appropriate commit in Falco.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

See inline comments.

**Does this PR introduce a user-facing change?**:

```release-note
new(falco): add gVisor support
```

### CODE REVIEW

File: userspace/falco/application.h [hunk 1]
```diff
@@ -193,6 +193,11 @@ class application {
 		return !m_options.trace_filename.empty();
 	}
 
+	inline bool is_gvisor_enabled() const
+	{
+		return m_state->inspector->is_gvisor();
```
Comment (LucaGuerra) [inline userspace/falco/application.h:198]: Note for the reviewer: the astute reader will have noticed that the boolean method `is_gvisor()` does not exist yet in falco libs, so I'm going to add it. I think it's the clean way to implement this check but if maintainers and contributors have better ideas I can change the implementation.

Comment (LucaGuerra) [reply userspace/falco/application.h:198]: Now it is implemented.

**Comment by FedeDP** [comment]
I think this should be rebased on master to fix integration tests!

**Comment by LucaGuerra** [comment]
* PR is rebased
* It now requires https://github.com/falcosecurity/libs/pull/429 because the interface changed.

**Comment by LucaGuerra** [comment]
Now it's rebased on top of libs that expose all the necessary functionality! :tada:

File: userspace/falco/app_cmdline_options.cpp [hunk 2]
```diff
@@ -162,6 +162,9 @@ void cmdline_options::define()
 		("disable-source",                "Disable a specific event source. Available event sources are: syscall or any source from a configured plugin with event sourcing capability. It can be passed multiple times. Can not disable all event sources.", cxxopts::value(disable_sources), "<event_source>")
 		("D",                             "Disable any rules with names having the substring <substring>. Can be specified multiple times. Can not be specified with -t.", cxxopts::value(disabled_rule_substrings), "<substring>")
 		("e",                             "Read the events from <events_file> in .scap format instead of tapping into live.", cxxopts::value(trace_filename), "<events_file>")
+		("g,gvisor-config",				  "Parse events from gVisor using the specified configuration file. A falco-compatible configuration file can be generated with --gvisor-generate-config and can be used for both runsc and Falco.", cxxopts::value(gvisor_config), "<gvisor_config>")
```
[the remark this answers is not in the document]
Comment (FedeDP) [reply userspace/falco/app_cmdline_options.cpp:165]: This LGTM! I am waiting for this part and we are ready to go IMO :)

[the remark this answers is not in the document]
Comment (leogr) [reply userspace/falco/app_cmdline_options.cpp:165]: Use the `MINIMAL_BUILD` pattern please :point_down: 
https://github.com/falcosecurity/falco/blob/master/userspace/falco/app_cmdline_options.cpp#L166

Options must be present only when Falco has been built with the gVisor support.

Furthermore, note that the gVisor support must be disabled when `MINIMAL_BUILD` is ON.

File: cmake/modules/falcosecurity-libs.cmake [hunk 3]
```diff
@@ -27,8 +27,8 @@ else()
   # In case you want to test against another falcosecurity/libs version (or branch, or commit) just pass the variable -
   # ie., `cmake -DFALCOSECURITY_LIBS_VERSION=dev ..`
   if(NOT FALCOSECURITY_LIBS_VERSION)
-    set(FALCOSECURITY_LIBS_VERSION "0.7.0-alpha2")
-    set(FALCOSECURITY_LIBS_CHECKSUM "SHA256=4a62071043aa50cdd29016c7dc041b401f8fa551271ec7b3ba01a537e5eb8bce")
+    set(FALCOSECURITY_LIBS_VERSION "a3b4b5bd586b491d329529d827d50443f98b240a")
```
Comment (leogr) [inline cmake/modules/falcosecurity-libs.cmake:30]: I'd suggest tagging a version in libs before merging this PR.

@FedeDP wdyt?

**Review by leogr — COMMENTED** [review]
This is awesome! :star_struck: 

Just left two comments (see below)

File: userspace/falco/app_actions/init_falco_engine.cpp [hunk 4]
```diff
@@ -43,6 +43,11 @@ void application::configure_output_format()
 		output_format = m_options.print_additional;
 		replace_container_info = false;
 	}
+	else if(m_options.gvisor_config != "")
+	{
+		output_format = "container=%container.id pid=%proc.vpid tid=%thread.vtid ";
```
Comment (jasondellaluce) [inline userspace/falco/app_actions/init_falco_engine.cpp:48]:
```suggestion
		output_format = "container=%container.id pid=%proc.vpid tid=%thread.vtid";
```

File: cmake/modules/falcosecurity-libs.cmake (hunk 3 above)
Comment (FedeDP) [reply cmake/modules/falcosecurity-libs.cmake:30]: Makes sense! :+1:

**Review by jasondellaluce — COMMENTED** [review]
Left you some comments Luca! Mostly nitpicking

File: userspace/falco/app_cmdline_options.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (LucaGuerra) [reply userspace/falco/app_cmdline_options.cpp:165]: I'm following that same pattern, reverting the PR to wip until I'm relatively happy with the result

File: userspace/falco/application.h [hunk 5]
```diff
@@ -193,6 +194,11 @@ class application {
...
```
[the remark this answers is not in the document]
Comment (LucaGuerra) [reply userspace/falco/application.h]: Good catch! I'm not completely convinced of both but now I see that the one just above it does exactly what you describe. I updated that to reflect this.

**Comment by LucaGuerra** [comment]
@leogr @jasondellaluce @FedeDP addressed review comments.
I also checked that it builds without gvisor support and in that case the CLI options related to it are not displayed at all.

**Comment by FedeDP** [comment]
>  I also checked that it builds without gvisor support and in that case the CLI options related to it are not displayed at all.

Did you also check that gvisor is actually not built inside libscap? I mean, that eg: you don't see the protobuf dep in libscap.

**Comment by FedeDP** [comment]
IMO this LGTM! I think that we can wait for https://github.com/falcosecurity/libs/pull/433 and https://github.com/falcosecurity/libs/pull/416 in libs and tag a new libs/driver release that can possibly be the final for the release.
WDYT? 
Do you want this merged before those 2 PRs?

**Comment by LucaGuerra** [comment]
@FedeDP  The PRs you mentioned are in and the versions have been updated in master. This can now be merged on top of master with the current RC libs versions. I just rebased the branch accordingly.

**Review by FedeDP — APPROVED** [review]
/approve

**Review by jasondellaluce — APPROVED** [review]

**Pull request merged**